### PR TITLE
DeepSeek V4 User Guide Update

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Adds XLML tests for DeepSeek V4 Flash and updates the `Run_DeepSeek.md` doc with V4 relevant instructions.

FIXES: b/509931363

# Tests

Ran `maxtext/tests/end_to_end/tpu/deepseek/v4-284b/2_test_deepseek.sh` with successful log output: https://cloudlogging.app.goo.gl/ox3QKj9JmgH2dSEx7

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by entrpn — APPROVED** [review]

**Review by parambole — COMMENTED** [review]
@snehalv2002 could you please update the PR title as this add test scripts and updates the user guides and not XLML tests which would be part of a different PR.

File: tests/end_to_end/tpu/deepseek/v4-284b/2_test_deepseek.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+# export UNSCANNED_CKPT_PATH=${BASE_OUTPUT_PATH}/unscanned/0/items
+# Use a hard-coded golden checkpoint, rather than checkpoints generated by Step 1 as it is not in daily test.
+SCANNED_CKPT_PATH=gs://maxtext-deepseek/deepseek4-284b/2026-07-24/scanned/0/items
+UNSCANNED_CKPT_PATH=gs://maxtext-deepseek/deepseek4-284b/2026-07-24/unscanned/0/items
+
+# Test whether the forward pass logits match the golden logits
+# default golden_logits_path=/deps/tests/assets/golden_logits/golden_data_{MODEL_NAME}.jsonl, copied from gs://maxtext-test-assets/golden_data_${MODEL_NAME}.jsonl
+GOLDEN_LOGITS_DISK_LOCATION="/deps/tests/assets/golden_logits/golden_data_${MODEL_NAME}.jsonl"
+if [ ! -f "${GOLDEN_LOGITS_DISK_LOCATION}" ]; then
+  GOLDEN_LOGITS_PATH="gs://snehalv-data/golden-logits-topk4/deepseek-ai/DeepSeek-V4-Flash/golden_data_deepseek4-284b_proper_512.jsonl"
```
Comment (shuningjin) [inline tests/end_to_end/tpu/deepseek/v4-284b/2_test_deepseek.sh]: Could you upload this to 
```
gs://maxtext-test-assets/golden_data_deepseek4-284b.jsonl
```

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 2]
```diff
@@ -26,67 +26,90 @@ DeepSeek is a novel family of open-weights sparse MoE models by DeepSeek AI. The
...
+    --hf_model_path=$BF16_LOCAL_PATH \
     --save_dtype=bfloat16
 ```
 
+
 ## Pre-training
-You can train from scratch to generate a new checkpoint. One example command to run pretraining with V3 on v5p-256.
+You can train from scratch or resume from a converted scanned checkpoint. The pre-training process and options are similar across all DeepSeek models.
+
+Here is an example command template. Export the variables `MODEL_NAME` (e.g., `deepseek3-671b`, `deepseek4-284b`), `TOKENIZER_PATH` (e.g., `deepseek-ai/DeepSeek-V3`, `deepseek-ai/DeepSeek-V4-Flash`), and customize other arguments such as `attention` (`flash` or `dot_product`), `opt_type` (`sgd` or `adamw`), and parallelism topology configs to match your model and setup:
```
Comment (shuningjin) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: add `muon`

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md
Comment (shuningjin) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: add `[V4](v4-284b)`

File: tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  base_output_directory=${BASE_OUTPUT_PATH}/scanned hf_access_token=$HF_TOKEN \
+  hardware=cpu skip_jax_distributed_system=True \
+  --hf_model_path=$BF16_LOCAL_PATH \
+  --lazy_load_tensors=False \
+  --simulated_cpu_devices_count=16 \
+  --save_dtype=bfloat16
+
+# unscanned
+# Note: Simulating CPU devices and disabling lazy load here also optimizes conversion performance.
+XLA_FLAGS="--xla_force_host_platform_device_count=16" python3 -m maxtext.checkpoint_conversion.to_maxtext src/maxtext/configs/base.yml \
```
Comment (shuningjin) [inline tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh]: Do we need this flag in command? The script sets the flag internally
https://github.com/AI-Hypercomputer/maxtext/blob/40aaa08f9a611f3781a785840da9c1ad22477c81/src/maxtext/checkpoint_conversion/to_maxtext.py#L1162

Do you see conversion speed difference due to jax device count? I thought it was only used at the very end when saving checkpoint.

**Review by shuningjin — COMMENTED** [review]
Thank you! Some minor comments.

File: tests/end_to_end/tpu/deepseek/v4-284b/2_test_deepseek.sh (hunk 1 above)
Comment (snehalv2002) [reply tests/end_to_end/tpu/deepseek/v4-284b/2_test_deepseek.sh]: done

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md (hunk 2 above)
Comment (snehalv2002) [reply tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: done

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md
Comment (snehalv2002) [reply tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: done

File: tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh (hunk 3 above)
Comment (snehalv2002) [reply tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh]: yeah good catch, that flag is not needed. removed it.

**Review by shuningjin — APPROVED** [review]
Thank you!

File: docs/reference/models/supported_models_and_architectures.md [hunk 4]
```diff
@@ -43,8 +43,8 @@ MaxText is an open-source, high-performance LLM framework written in Python/JAX.
 
 ### DeepSeek
 
-- **Variants**: V2 (16B, 236B), V3 (671B) / R1 / V3.1, V3.2 (671B)
-- **Notes**: MLA; shared/finer-grained experts; MTP; YaRN-style scaling; DeepSeek Sparse Attention (V3.2).
+- **Variants**: V2 (16B, 236B), V3 (671B) / R1 / V3.1, V3.2 (671B), V4 Flash (284B)
+- **Notes**: MLA; shared/finer-grained experts; MTP; YaRN-style scaling; DeepSeek Sparse Attention (V3.2); Heavily compressed attention + compressed sparse attention + manifold-constrained hyperconnections (V4).
```
Comment (parambole) [inline docs/reference/models/supported_models_and_architectures.md]: @snehalv2002 - Please also include hashrouting.

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 5]
```diff
@@ -304,11 +327,57 @@ python3 -m tests.utils.forward_pass_logit_checker \
...
+    ici_expert_parallelism=-1 \
+    weight_dtype=bfloat16 \
+    dtype=bfloat16 \
+    activations_in_float32=false \
+    matmul_precision=highest \
+    float32_logits=false \
+    float32_qk_product=false \
+    override_model_config=True \
+    indexer_topk=4 \
+    enable_nnx=false \
```
Comment (parambole) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: @snehalv2002 - Please add a note around nnx here.

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 6]
```diff
@@ -26,67 +26,90 @@ DeepSeek is a novel family of open-weights sparse MoE models by DeepSeek AI. The
...
+    model_name=${MODEL_NAME} \
+    ici_fsdp_parallelism=-1 \
     steps=5 \
     max_target_length=1024 \
     async_checkpointing=false \
     tokenizer_type=huggingface \
-    tokenizer_path=deepseek-ai/DeepSeek-V3 \
-    attention=flash \
+    tokenizer_path=${TOKENIZER_PATH} \
+    attention=flash \ # DSV4 does not yet support flash attention, need to use dot_product for now.
```
Comment (parambole) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]:
```+ # Note: DSV4 does not yet support flash attention; use dot_product for now.
  tokenizer_path=${TOKENIZER_PATH} \
-   attention=flash \ # DSV4 does not yet support flash attention, need to use dot_product for now.
+   attention=dot_product \
    dtype=bfloat16 \
```

The current way will break the command.

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 7]
```diff
@@ -26,67 +26,90 @@ DeepSeek is a novel family of open-weights sparse MoE models by DeepSeek AI. The
...
 * The provided TPU configurations are examples and not mandatory.
 * For V3.1 & R1, use existing V3 671B model configurations, as it shares the same architecture.
 
 ## Checkpoint conversion
-To get started, follow the instructions at HuggingFace ([V3](https://huggingface.co/deepseek-ai/DeepSeek-V3), [V2-Lite](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite)) to download the model. Currently for V3, V3.1, and R1, it uses mixed precision fp8 & bf16 weights. To convert all FP8 weights to BF16, use the script [here](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/utils/ckpt_scripts/deepseek_fp8_to_bf16.py). Once downloaded and converted to BF16:
+To get started, follow the instructions at HuggingFace ([V3](https://huggingface.co/deepseek-ai/DeepSeek-V3), [V2-Lite](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite)) to download the model. Currently for V3, V3.1, and R1, it uses mixed precision fp8 & bf16 weights. To convert all FP8 weights to BF16, use the script [here](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/utils/ckpt_scripts/deepseek_dequantize.py). Once downloaded and converted to BF16:
```
Comment (parambole) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: @snehalv2002 does this location exist ?

alt: src/maxtext/checkpoint_conversion/standalone_scripts/deepseek_dequantize.py?

File: tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+if [ -z "${CKPT_DISK_LOCATION}" ]; then
+  export BF16_HF_BUCKET=${BF16_HF_PATH}
+  gcloud storage cp -r ${BF16_HF_BUCKET} /tmp
+  export BF16_LOCAL_PATH=/tmp/hf-bf16
+else
+  export BF16_LOCAL_PATH=${CKPT_DISK_LOCATION}
+fi
+
+# scanned
+# Note: The following command uses XLA_FLAGS and --simulated_cpu_devices_count to simulate
```
Comment (parambole) [inline tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh]:
```- # Note: The following command uses XLA_FLAGS and --simulated_cpu_devices_count to simulate
+ # Note: The following command uses --simulated_cpu_devices_count to simulate
```
As there are no XLA flags.

File: tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh [hunk 9]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+# This file is documentation for how to get started with DeepSeek v4.
+
+# This file runs Step 1 on CPU.
+# 1. Convert the HuggingFace checkpoint (bf16) to MaxText-compatible checkpoint (bf16): 
+#    Scanned format is better for training; unscanned format is better for decoding.
+# 2. Run logit check and pre-training.
```
Comment (parambole) [inline tests/end_to_end/tpu/deepseek/v4-284b/1_test_deepseek.sh]: @snehalv2002 - Step 2 doesn't exist in this file we should remove this comment.

**Review by parambole — APPROVED** [review]
@snehalv2002 - I have left a few nits. Please resolve them before merging. 

LGTM.

**Review by aireenmei — APPROVED** [review]
